tools/ghsa/convert_ghsa.py

```diff
@@ -0,0 +1,277 @@
+
+        spec = GhsaVersionSpec(operator, version)
+
+        if operator == '=':
+            parsed_range.exact = spec
+        elif operator in ('>=', '>'):
+            parsed_range.lower = spec
+        elif operator in ('<=', '<'):
+            parsed_range.upper = spec
+        else:
+            raise ValueError(f'Unknown operator "{operator}"')
+
+    if parsed_range.exact and (parsed_range.lower or parsed_range.upper):
+        raise ValueError('Range with both exact and lower/upper bounds: '
+                         f'"{ghsa_range}"')
+
+    return parsed_range
+
+
+def convert_file(input_path: str, output_path: str):
+    """Converts `input_path` from GHSA JSON and output OSV JSON at
+    `output_path`."""
+    with open(input_path) as handle:
+        ghsa = json.load(handle)
+
+    osv = convert(ghsa)
+    with open(output_path, 'w') as handle:
+        handle.write(json.dumps(osv, indent=2))
+
+
+def convert_reference(reference: Dict[str, str]):
+    """Converts a GHSA reference to an OSV reference."""
+    ref_type = 'WEB'
+
+    if 'github.com/advisories/' in reference['url']:
+        ref_type = 'ADVISORY'
+
+    if 'nvd.nist.gov/vuln/detail/' in reference['url']:
+        ref_type = 'ADVISORY'
+
+    return {
+        'type': ref_type,
+        'url': reference['url'],
+    }
+
+
+def convert(ghsa: Dict[str, Any]):
+    """Converts a GHSA entry to an OSV entry."""
+    osv = {
+        'id':
+        ghsa['ghsaId'],
+        'aliases': [
+            val['value'] for val in ghsa['identifiers']
+            if val['value'] != ghsa['ghsaId']
+        ],
+        'published':
+        ghsa['publishedAt'],
+        'modified':
+        ghsa['updatedAt'],
+    }
+
+    # Split up the dict assignments to preserve order of date related fields.
+    withdrawn = ghsa.get('withdrawnAt')
+    if withdrawn:
+        osv['withdrawn'] = withdrawn
+
+    osv.update({
+        'summary':
+        ghsa['summary'],
+        'description':
+        ghsa['description'],
+        'references': [convert_reference(ref) for ref in ghsa['references']]
+    })
+
+    osv['affected'] = get_affected(ghsa)
+    return osv
+
+
+def get_affected(ghsa: Dict[str, Any]):
+    """Converts the GHSA entry into an OSV "affected" entry."""
+    package_to_vulns = {}
+
+    # Group vulnerabilities by (ecosystem, package).
+    for vuln in ghsa.get('vulnerabilities', []).get('nodes', []):
+        package = vuln['package']
+        mapped_ecosystem = ECOSYSTEM_MAP[package['ecosystem']]
+        package_to_vulns.setdefault((mapped_ecosystem, package['name']),
+                                    []).append(vuln)
+
+    cvss = ghsa.get('cvss', {})
+    cwes = ghsa.get('cwes', {}).get('nodes', [])
+
+    # Convert the grouped vulnerabilities in OSV range structures.
+    affected = []
+    for (ecosystem, name), vulns in package_to_vulns.items():
+        current = {
+            'package': {
+                'ecosystem': ecosystem,
+                'name': name,
```

Comment:
I assume GHSA uses same normalization rules as OSV?

Reply:
Thanks for catching this! Looks like for PyPI at least, GHSA does not perform normalization. I've fixed this.